backend: Add request parameter to `json_success`.

laurynmm opened the pull request:
Refactors `json_success` to accept a request parameter, which is a prerequisite for making the `ignored_parameters_unsupported` functionality available on all API endpoints. See [this CZO chat](https://chat.zulip.org/#narrow/stream/3-backend/topic/ignored_parameters_unsupported) for more context.

The first 4 commits are preparatory steps:
- Passing down the request to a few functions in views that did not previously do so.
- Removing an occurrence of `json_success` in a function called from a queue processor.
- Removing a custom linter rule that is no longer useful.

The 5th commit:
- Adjusts the signature of `json_success` to be `json_success(request, ...)`, but doesn't use the new parameter.
- Adds said parameter to all occurrences of `json_success' in the repository.

laurynmm commented on zerver/tests/test_subs.py:
```diff
@@ -3514,6 +3514,7 @@ def test_compose_views_rollback(self) -> None:
         user_profile.full_name = "Hamlet"
         user_profile.save()
+        request = HttpRequest()
 
```
I wasn't sure what to do in this case because what's being tested is the Django rollback, which should not make these changes. But I can imagine this change being an issue when `json_success` is updated to use the request data.

timabbott replied:
I think we have a standard class, `HostRequestMock`, for this purpose.

laurynmm commented on zerver/views/user_groups.py:
```diff
@@ -103,7 +103,7 @@ def update_user_group_backend(
     ]
-    return compose_views(thunks)
+    return compose_views(request, thunks)
 
```
Also, since `compose_views` is being used by two views should / could it be moved to be in a more general spot for helper functions?

timabbott commented on zerver/views/muting.py:
```diff
@@ -44,7 +45,11 @@ def mute_topic(
+    user_profile: UserProfile,
+    stream_id: Optional[int],
+    stream_name: Optional[str],
+    topic_name: str,
```
I think for these view helpers, it'd be better to just change the interface for this function to just return None (and have the caller do the `json_success`; it only fails via throwing an exception anyway)

timabbott commented on zerver/views/message_send.py:
```diff
@@ -285,6 +286,7 @@ def send_message_backend(
 
     if (delivery_type == "send_later" or delivery_type == "remind") and defer_until is not None:
         return handle_deferred_message(
+            request,
```
Probably same with this one; the view helper could return `str(deliver_at_usertz)` and then this could do the `json_success` part.

timabbott commented on zerver/views/streams.py:
```diff
@@ -374,10 +374,10 @@ def update_subscriptions_backend(
 
 
-def compose_views(thunks: List[Callable[[], HttpResponse]]) -> HttpResponse:
+def compose_views(request: HttpRequest, thunks: List[Callable[[], HttpResponse]]) -> HttpResponse:
```
I think we can also avoid this API change by having this just return `json_dict`, and have the caller(s) do the `json_success` wrapping bit.

timabbott commented on zerver/lib/error_notify.py:
```diff
@@ -194,11 +192,10 @@ def email_server_error(report: Dict[str, Any]) -> None:
         notify_server_error(report)
     else:
         raise JsonableError(_("Invalid type parameter"))
-    return json_success()
```
Yes, that's correct, but I think it still makes sense to convert it to an `AssertionError`, because `JsonableError` isn't appropriate for this context.

timabbott commented on tools/linter_lib/custom_check.py:
```diff
@@ -325,12 +325,6 @@
-            "description": "Use json_success() to return nothing",
-            "good_lines": ["return json_success()"],
-            "bad_lines": ["return json_success({})"],
-        },
```
I think we could reasonably change this to ban `json_success(request, {})` instead, but I think that realistically, nobody is going to type that, so yeah, we can remove this.

laurynmm replied:
Agreed. Interestingly, I did find at least one instance of `data={}` and a few instances of `{"type": "success", "msg": ""}` in the search/replace commit that I removed, perhaps other small things that were missed in previous refactors.

timabbott commented:
This is looking good @laurynmm!  I posted a batch of comments on how we might better address the corner cases covered in the prep commits.  We should also be sure to adjust the references in documentation:
```
$ git grep json_success docs/
docs/testing/testing-with-django.md:- assert_json_success
docs/tutorials/life-of-a-request.md:`json_success(data=<some python object which can be converted to a JSON string>)`.
tabbott@coset:~/zulip$ git grep json_success templates/
templates/zerver/api/incoming-webhooks-walkthrough.md:from zerver.lib.response import json_success
templates/zerver/api/incoming-webhooks-walkthrough.md:    return json_success()
templates/zerver/api/incoming-webhooks-walkthrough.md:`json_success()`.
templates/zerver/api/incoming-webhooks-walkthrough.md:`send_webhook_payload`, which checks the result with `assert_json_success`.
```

laurynmm commented:
@timabbott Thanks for the detailed review; it helped me learn a quite a bit about how to go about these kinds of tasks.

I've gone back through and updated the preparatory commits re: your comments and added a separate commit for the documentation updates. Let me know if anything looks off or still needs work.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -3512,6 +3513,7 @@ def test_compose_views_rollback(self) -> None:
         user_profile = self.example_user("hamlet")
         user_profile.full_name = "Hamlet"
         user_profile.save()
+        request = HostRequestMock(user_profile=user_profile)
```
The changes in this file are no longer needed; I'll drop them while merging.

timabbott commented on docs/tutorials/life-of-a-request.md:
```diff
@@ -203,8 +203,13 @@ in a
+a successfully handled request, we use `json_success(request, data)`.
+
+The request argument is a
+[HTTP request](https://docs.djangoproject.com/en/3.2/ref/request-response/)
```
Let's spell this `HttpRequest`; since that's the name of the Django class, and it's helpful to be precise where we can.

I'll do a pass on this paragraph while thinking about this.

timabbott replied:
Added this as a new bonus commit before merging.

timabbott commented:
This is great, merged with the small changes noted above, thanks @laurynmm!  

Congratulations on completing your first big refactor :).  I think there are a few next steps to pursue in parallel:
* Try moving/copying this block to run in every `json_success` and `JsonableError` call, and write a test suite for this feature (Perhaps in `test_decorators.py`, using an artificial view defined just in that file?  Not sure what's cleanest), and see what breaks / needs updating.  One step we'll want to do here is basically add a temporary assertion that this never has anything in it HTTP requests in the test suite, and to see what tests fail with that assertion (this should be the case anywhere that we are parsing parameters not using `REQ`, I think).  That will likely spawn a series of discussions/refactors/etc. to resolve such situations into a state we're happy with.
* Bump the `#api design` thread about this with the question of whether we want `ignored_parameters_unsupported` to be present unconditionally, or only when there are some.  I'm most tempted by preserving the current model, where we don't mention parameters that are unused, since that way if you're manually inspecting HTTP responses from Zulip, you only see this property if you have a potential bug.  But it's worth discussion since there's arguments both way.

```
    # TODO: Extract `ignored_parameters_unsupported` to be a common feature of the REQ framework.        
    from zerver.lib.request import RequestNotes                                                          
                                                                                                         
    request_notes = RequestNotes.get_notes(request)                                                      
    for req_var in request.POST:                                                                         
        if req_var not in request_notes.processed_parameters:                                            
            request_notes.ignored_parameters.add(req_var)                                                
                                                                                                         
    result: Dict[str, Any] = {}                                                                          
    if len(request_notes.ignored_parameters) > 0:                                                        
        result["ignored_parameters_unsupported"] = list(request_notes.ignored_parameters)                
                                                                                                         
    return json_success(request, data=result)                                                            
```

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -3516,7 +3516,7 @@ def test_compose_views_rollback(self) -> None:
             user_profile.full_name = "Should not be committed"
             user_profile.save()
-            return json_success()
+            return json_success(request)
```
I think I handled this wrong, resulting in CI failures in main -- I think we wanted to remove the `json_success` from this function altogether when changing the `compose_views` API.

laurynmm replied:
@timabbott Yeah, the `thunks` being called in `compose_views` are view functions that return `json_success` when everything works (and that rollback the database for any changes made / raise `JsonableError` when there's an issue). So, I'm a bit stuck on how we could eliminate the `json_success` from those single add / delete views.

We could put the rollback and `thunks` loop back into the two view functions using `compose_views`. It looks like @showell  refactored this to be a separate function in `streams.py` about 6 years ago and it was added to `user_groups.py` about a year after that. If we made that change then we'd need to test the rollback within the two views vs testing the helper. I guess can see doing that if we feel like these types of endpoints won't be common / increasing, but there might also be considerations I'm not aware of as well.

In the use case for `streams.py`, the `thunk` views are also actively used in other rest paths in `zproject.urls.py`. Also, that [update subscriptions endpoint using `compose_views`](https://zulip.com/api/update-subscriptions) is documented in the OpenAPI, but is not in the api docs sidebar so not easily find-able for folks looking at the documentation site. So maybe it's not an endpoint we're encouraging clients to use? And eventually eliminating?

In the use case for `user_groups.py`, the `thunk` views are currently only used for the [one rest path](https://zulip.com/api/update-user-group). That endpoint is documented and in the sidebar list in this case. This seems like an endpoint that is encouraged for use, but also would be easier to refactor since the add / delete view functions aren't being used outside of the one endpoint.

Does this merit a CZO topic conversation? I can move this all there if that makes the most sense. Either to backend or api_design ...

timabbott replied:
I think #backend is probably fine for this, since it's an implementation detail.  Seems worth raising there; comments on closed GitHub issues tend to get lost :).

But yeah, it's very possible we want to deprecate endpoints using this pattern; it's a bit of a weird one dating to very early in Zulip's history.
